# General statistics

## Sample statistics

### All

```{r samples_stats, warning=FALSE, comments="", message=FALSE}
read_tsv("data/sample.tsv") %>%
  summarise(specimens=n_distinct(specimen_id),
            species=n_distinct(specimen_species),
            orders=n_distinct(specimen_order),
            families=n_distinct(specimen_family)) %>% 
  tt()
```

### Subset

Faecal and cloacal swab samples employed in the study.

```{r samples_stats_subset, warning=FALSE, comments="", message=FALSE}
inner_join(read_tsv("data/sample.tsv"),
          read_tsv("data/extraction.tsv"),
          by="sample_id") %>% 
  filter(sample_type %in% c("Faecal", "Anal/cloacal swab")) %>%
  summarise(specimens=n_distinct(specimen_id),
            species=n_distinct(specimen_species),
            orders=n_distinct(specimen_order),
            families=n_distinct(specimen_family),
            swabs=n_distinct(sample_id[sample_type == "Anal/cloacal swab"]),
            faeces=n_distinct(sample_id[sample_type == "Faecal"])) %>% 
  tt()
```

### Origin of samples (Figure S1)

```{r sample_map, message=F, warning=F, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
read_tsv("data/sample.tsv") %>%
  #subset columns
  dplyr::select(
    sample_id,
    specimen_species,
    specimen_order,
    specimen_class,
    capture_latitude,
    capture_longitude
  ) %>%
  #Add jitter to points
  mutate(
    capture_latitude_jitter=capture_latitude+rnorm(length(capture_latitude), mean=0, sd=0.5),
    capture_longitude_jitter=capture_longitude+rnorm(length(capture_longitude), mean=0, sd=0.5),
  ) %>%
  #Plot map  
  ggplot(.) +
    geom_map(
      data=map_data("world"),
      map = map_data("world"),
      aes(long, lat, map_id=region),
      color = "white", fill = "#cccccc", size = 0.2
    ) +
    geom_point(
      aes(x=capture_longitude_jitter,y=capture_latitude_jitter, color=specimen_order),
      alpha=0.5, size=0.5, shape=16) +
    labs(color="Taxonomic order") +
    theme_minimal() +
    theme(
      axis.title.x=element_blank(),
      axis.title.y=element_blank(),
      legend.position = "bottom")
```

## Data statistics

### Total data

```{r total_data, warning=FALSE, comments="", message=FALSE}
left_join(read_tsv("data/preprocessing.tsv"),
          read_tsv("data/sample.tsv"),
          by="sample_id") %>%
  filter(sample_type %in% c("Faecal", "Anal/cloacal swab")) %>%
  mutate(bases_pre_fastp = bases_pre_fastp / 1000000000)  %>% #convert bases to gigabases (GB)
  summarise(
    total= sum(bases_pre_fastp, na.rm = TRUE),
    swabs=sum(bases_pre_fastp[sample_type == "Anal/cloacal swab"], na.rm = TRUE),
    faeces=sum(bases_pre_fastp[sample_type == "Faecal"], na.rm = TRUE),
    swabs_n=n_distinct(preprocessing_id[sample_type == "Anal/cloacal swab"], na.rm = TRUE),
    faeces_n=n_distinct(preprocessing_id[sample_type == "Faecal"], na.rm = TRUE),
    mean= mean(bases_pre_fastp, na.rm = TRUE),
    sd = sd(bases_pre_fastp, na.rm = TRUE),
    median=median(bases_pre_fastp, na.rm = TRUE),
    IQR=IQR(bases_pre_fastp, na.rm = TRUE)
  ) %>%
  tt()
```

### Quality-filtered data

```{r quality_filtered_data, warning=FALSE, comments="", message=FALSE}
read_tsv("data/preprocessing.tsv") %>%
  mutate(bases_post_fastp = bases_post_fastp / 1000000000)  %>% #convert bases to gigabases (GB)
  summarise(
    total= sum(bases_post_fastp, na.rm = TRUE),
    mean= mean(bases_post_fastp, na.rm = TRUE),
    sd = sd(bases_post_fastp, na.rm = TRUE),
    median=median(bases_post_fastp, na.rm = TRUE),
    IQR=IQR(bases_post_fastp, na.rm = TRUE)
  ) %>%
  tt()
```

### Host genomic data

```{r host_data, warning=FALSE, comments="", message=FALSE}
read_tsv("data/preprocessing.tsv") %>%
  mutate(host_bases = host_bases / 1000000000)  %>% #convert bases to gigabases (GB)
  summarise(
    total= sum(host_bases, na.rm = TRUE),
    mean= mean(host_bases, na.rm = TRUE),
    sd = sd(host_bases, na.rm = TRUE),
    median=median(host_bases, na.rm = TRUE),
    IQR=IQR(host_bases, na.rm = TRUE)
  ) %>%
  tt()
```

### Metagenomic data

```{r metagenomic_data, warning=FALSE, comments="", message=FALSE}
read_tsv("data/preprocessing.tsv") %>%
  mutate(metagenomic_bases = metagenomic_bases / 1000000000)  %>% #convert bases to gigabases (GB)
  summarise(
    total= sum(metagenomic_bases, na.rm = TRUE),
    mean= mean(metagenomic_bases, na.rm = TRUE),
    sd = sd(metagenomic_bases, na.rm = TRUE),
    median=median(metagenomic_bases, na.rm = TRUE),
    IQR=IQR(metagenomic_bases, na.rm = TRUE)
  ) %>%
  tt()
```

### Assemblies 

```{r number_of_assemblies, warning=FALSE, comments="", message=FALSE}
left_join(read_tsv("data/assembly.tsv"),
          read_tsv("data/preprocessing.tsv"),
          by="preprocessing_id") %>%
  left_join(read_tsv("data/sample.tsv"),by="sample_id") %>% 
  group_by(assembly_type) %>% 
  summarise(assembly_n=n_distinct(assembly_id),
            swabs_n=n_distinct(assembly_id[sample_type == "Anal/cloacal swab"], na.rm = TRUE),
    faeces_n=n_distinct(assembly_id[sample_type == "Faecal"], na.rm = TRUE),
    swabs_size=sum(assembly_length[sample_type == "Anal/cloacal swab"], na.rm = TRUE),
    faeces_size=sum(assembly_length[sample_type == "Faecal"], na.rm = TRUE)) %>%
  tt()
```

### MAGs

```{r number_of_mags, warning=FALSE, comments="", message=FALSE}
left_join(read_tsv("data/mag.tsv"),
          read_tsv("data/assembly.tsv"),
          by="assembly_id") %>%
    left_join(read_tsv("data/preprocessing.tsv"),by="preprocessing_id") %>% 
    left_join(read_tsv("data/sample.tsv"),by="sample_id") %>% 
    dplyr::select(mag_id,mag_phylum,mag_completeness,mag_contamination,sample_type)%>% 
    unique() %>% 
    summarise(number=n(),
            swabs_n=n_distinct(mag_id[sample_type == "Anal/cloacal swab"], na.rm = TRUE),
            faeces_n=n_distinct(mag_id[sample_type == "Faecal"], na.rm = TRUE),
            phylums=n_distinct(mag_phylum),
            completeness=mean(mag_completeness),
            contamination=mean(mag_contamination)) %>% 
  tt()
```

